compiler/rustc_mir_transform/src/jump_threading.rs

```diff
@@ -0,0 +1,759 @@
+            let mut new_involved = Vec::new();
+            for &(to_index, in_to_index) in &self.involving_tos[current] {
+                // That TO has already been applied, do nothing.
+                if to_index <= index {
+                    continue;
+                }
+
+                let other_to = &mut self.opportunities[to_index];
+                if other_to.chain.get(in_to_index) != Some(&current) {
+                    continue;
+                }
+                let s = other_to.chain.get_mut(in_to_index + 1).unwrap_or(&mut other_to.target);
+                if *s == succ {
+                    // `other_to` references the `current -> succ` edge, so replace `succ`.
+                    *s = new_succ;
+                    new_involved.push((to_index, in_to_index + 1));
+                }
+            }
+
+            // The TOs that we just updated now reference `new_succ`. Update `involving_tos`
+            // in case we need to duplicate an edge starting at `new_succ` later.
+            let _new_succ = self.involving_tos.push(new_involved);
+            debug_assert_eq!(new_succ, _new_succ);
+
+            current = new_succ;
+        }
+
+        let current = &mut basic_blocks[current];
+        self.update_predecessor_count(current.terminator(), Update::Decr);
+        current.terminator_mut().kind = TerminatorKind::Goto { target: op_target };
+        self.predecessors[op_target] += 1;
+    }
+
+    fn update_predecessor_count(&mut self, terminator: &Terminator<'_>, incr: Update) {
+        match incr {
+            Update::Incr => {
+                for s in terminator.successors() {
+                    self.predecessors[s] += 1;
+                }
+            }
+            Update::Decr => {
+                for s in terminator.successors() {
+                    self.predecessors[s] -= 1;
+                }
+            }
+        }
+    }
+}
+
+fn predecessor_count(body: &Body<'_>) -> IndexVec<BasicBlock, usize> {
+    let mut predecessors: IndexVec<_, _> =
+        body.basic_blocks.predecessors().iter().map(|ps| ps.len()).collect();
+    predecessors[START_BLOCK] += 1; // Account for the implicit entry edge.
+    predecessors
+}
+
+enum Update {
+    Incr,
+    Decr,
+}
+
+/// Compute the set of loop headers in the given body. We define a loop header as a block which has
+/// at least a predecessor which it dominates. This definition is only correct for reducible CFGs.
+/// But if the CFG is already irreducible, there is no point in trying much harder.
+/// is already irreducible.
+fn loop_headers(body: &Body<'_>) -> BitSet<BasicBlock> {
+    let mut loop_headers = BitSet::new_empty(body.basic_blocks.len());
+    let dominators = body.basic_blocks.dominators();
+    // Only visit reachable blocks.
+    for (bb, bbdata) in traversal::preorder(body) {
+        for succ in bbdata.terminator().successors() {
+            if dominators.dominates(succ, bb) {
+                loop_headers.insert(succ);
```

Comment:
This was wrong. Added a correction commit.

Reply:
So... the effect of this change is two-fold (when succ dominates bb): 1. it avoids adding `bb` to the loop headers, and 2. it adds `succ` to the loop headers.

I see essentially only one effect on the tests though (where that one effect is duplicated in two twos) -- and I *think* the effect demonstrated by the tests is due to the addition of adding `succ` to the loop headers (since that cancels out a threading opportunity, which I believe is what is happening in the panic-abort and panic-unwind test)

So, does this imply that there is some missing test coverage, i.e. a test that demonstrates how removing `bb` from loop headers *enables* some jump threading to proceed?

Reply:
(I'm still going to r+ the PR as is, because the missing test, I think, only represents a failure to demonstrate a newly present optimization opportunity, which while nice to have in tests, is not as important as tests that are catching potential soundness issues in the transformation.)

Reply:
This correction is a change in the definition. The restriction on loop headers is purely heuristic, and has no impact on soundness. LLVM does not like irreducible loops, so we avoid creating them.
The DFA test means precisely to capture this, but was overlooked. With a845bac6ba434483b483c0561011227c1eff9daf it should not break silently any more.